sql-plugin/src/main/scala/com/nvidia/spark/rapids/RmmRapidsRetryIterator.scala

```diff
@@ -621,7 +627,8 @@ object RmmRapidsRetryIterator extends Logging {
       var lastException: Throwable = null
       var firstAttempt: Boolean = true
       var result: Option[K] = None
-      var doSplit = false
+      var doSplitForStateMachineReasons = false
```

Comment:
Could here use a single variable of SplitReason instead ? I think this will make this change smaller.
`var splitReason = SplitReason.NONE`, and trigger split when it is not a NONE.

Reply:
but the idea of NONE might be confusing for cases where it is actually able to split. 

We are more thinking about withRetryNoSplit case, but there're also cases where ColumnSizeOverflow can be handled by split and retry, in a withRetry context

Reply:
I mean we can have ,for example `GPU_OOM, CPU_OOM, COLUMN_OVER_FLOW, NONE`.
`NONE` here just is for the initial value, like false for a boolean.
Then we re-assign splitReason to other values accordingly, and trigger split when it is not NONE.

`doSplitForStateMachineReasons = true` can be replaced by `splitReason =GPU_OOM ` or `splitReason =CPU_OOM `.
While `splitReason=COLUMN_OVER_FLOW` is for `doSplitForNonStateMachineReasons = true`.

Two boolean variables can represent 4 states, so let splitReason have 4 values respectively.

Reply:
I may miss something, however 3 variables here are more error-prone than a single one. And I am still not fully get the relationship among them.

Reply:
same with in other thread,  I'm not sure if we want so many enums

Reply:
what do you mean by 3 variables ?

Reply:
i'll try to add some modifications to reduce variable num

Reply:
hi @firestarman  thx for the adivse, I have reduced the number of variables